core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/store.go

```diff
@@ -104,57 +132,87 @@ func (u *upkeepStateStore) Close() error {
 // SelectByWorkIDs returns the current state of the upkeep for the provided ids.
 // If an id is not found, the state is returned as StateUnknown.
 // We first check the cache, and if any ids are missing, we fetch them from the scanner.
-// TODO: fetch from DB
 func (u *upkeepStateStore) SelectByWorkIDsInRange(ctx context.Context, start, end int64, workIDs ...string) ([]ocr2keepers.UpkeepState, error) {
-	states, ok := u.selectFromCache(workIDs...)
+	hWorkIDs := make([]common.Hash, len(workIDs))
+	for i, workID := range workIDs {
+		hWorkIDs[i] = common.HexToHash(workID)
+	}
+
+	states, ok := u.selectFromCache(hWorkIDs...)
 	if ok {
 		// all ids were found in the cache
 		return states, nil
 	}
+
+	// fetch values from chain to populate the cache with missing values
 	if err := u.fetchPerformed(ctx, start, end); err != nil {
 		return nil, err
 	}
-	states, _ = u.selectFromCache(workIDs...)
+
+	idsWithUnknownState := []common.Hash{}
```

Comment:
extract lines 152-164 into it's own function (`fetchFromDB`), keep the main function clean